hibernate-core/src/main/java/org/hibernate/id/CompositeNestedGeneratedValueGenerator.java

```diff
@@ -107,11 +128,29 @@ public void addGeneratedValuePlan(GenerationPlan plan) {
 	public Object generate(SharedSessionContractImplementor session, Object object) throws HibernateException {
 		final Object context = generationContextLocator.locateGenerationContext( session, object );
 
+		final List<Object> generatedValues = compositeType.isMutable() ?
+				null :
+				new ArrayList<>( generationPlans.size() );
 		for ( GenerationPlan generationPlan : generationPlans ) {
-			generationPlan.execute( session, object, context );
+			final Object generated = generationPlan.execute( session, object );
+			if ( generatedValues != null ) {
+				generatedValues.add( generated );
+			}
+			else {
+				generationPlan.getInjector().set( context, generated );
+			}
 		}
 
-		return context;
+		if ( generatedValues != null) {
+			final Object[] values = compositeType.getPropertyValues( context );
+			for ( int i = 0; i < generatedValues.size(); i++ ) {
+				values[generationPlans.get( i ).getPropertyIndex()] = generatedValues.get( i );
+			}
+			return compositeType.replacePropertyValues( context, values, session );
+		}
+		else {
+			return context;
+		}
```

Comment:
@beikov nested composite value generation works with java records too, I used `CompositeType#replacePropertyValues` like you suggested.